src/typeahead/typeahead.ts

```diff
@@ -0,0 +1,132 @@
+    '(keydown.enter)': '_onKeyEnter()',
+    '(keydown.esc)': '_closePopup()'
+  }
+})
+export class NgbTypeahead implements OnInit {
+  private _popupService: PopupService<NgbTypeaheadWindow>;
+  private _positioning = new Positioning();
+  private _windowRef: ComponentRef<NgbTypeaheadWindow>;
+
+  /**
+   * A function to transform the provided observable text into the array of results
+   */
+  @Input() ngbTypeahead: (text: Observable<string>) => Observable<any[]>;
+
+  /**
+   * A function to format a given result before display. This function should return a formatted string without any
+   * HTML markup.
+   */
+  @Input() resultFormatter: (value) => string;
+
+  /**
+   * A template to display a matching result.
+   */
+  @Input() resultTemplate: TemplateRef<ResultTplCtx>;
+
+  constructor(
+      private _model: NgModel, private _elementRef: ElementRef, private _viewContainerRef: ViewContainerRef,
+      private _renderer: Renderer, private _injector: Injector, componentFactoryResolver: ComponentFactoryResolver) {
+    this._popupService = new PopupService<NgbTypeaheadWindow>(
+        NgbTypeaheadWindow, _injector, _viewContainerRef, _renderer, componentFactoryResolver);
+  }
+
+  ngOnInit() {
+    this._model.control.valueChanges.filter((value) => isString(value)).let (this.ngbTypeahead).subscribe((results) => {
+
+      if (!results || results.length === 0) {
+        this._closePopup();
+      } else {
+        this._openPopup();
+        this._windowRef.instance.results = results;
+        this._windowRef.instance.term = this._model.value;
+        if (this.resultFormatter) {
+          this._windowRef.instance.formatter = this.resultFormatter;
+        }
+        if (this.resultTemplate) {
+          this._windowRef.instance.resultTemplate = this.resultTemplate;
+        }
+      }
+    });
+  }
+
+  private _onKeyEnter() {
+    if (this._windowRef) {
+      const result = this._windowRef.instance.getActive();
+      this._selectResult(result);
+    }
+  }
+
+  private _onKeyUp(event: KeyboardEvent) {
+    if (this._windowRef) {
+      event.preventDefault();
+      this._windowRef.instance.prev();
+    }
+  }
+
+  private _onKeyDown(event: KeyboardEvent) {
+    if (this._windowRef) {
+      event.preventDefault();
+      this._windowRef.instance.next();
+    }
+  }
+
+  private _selectResult(result: any) {
+    this._model.viewToModelUpdate(result);
+    this._model.valueAccessor.writeValue(result);
+    this._closePopup();
+  }
+
+  private _closePopup() {
+    this._popupService.close();
+    this._windowRef = null;
+  }
+
+  private _openPopup() {
+    if (!this._windowRef) {
+      this._windowRef = this._popupService.open();
+      this._windowRef.instance.selectEvent.subscribe((result: any) => this._selectResult(result));
+
+      // position
+      const targetPosition = this._positioning.positionElements(
+          this._elementRef.nativeElement, this._windowRef.location.nativeElement, 'bottom', false);
```

Comment:
Shouldn't we use `bottom-left` instead of `bottom` so the popup is left-aligned instead of being centered?